src/xml.rs

```diff
@@ -0,0 +1,89 @@
+use std::ffi::CString;
+use std::fmt::{Debug, Display, Formatter};
+use std::ptr::NonNull;
+use std::str::FromStr;
+
+use gdal_sys::{
+    CPLCloneXMLTree, CPLDestroyXMLNode, CPLErr, CPLParseXMLString, CPLSerializeXMLTree, CPLXMLNode,
+};
+
+use crate::utils::_last_cpl_err;
+
+/// An XML node, as captured from GDAL serialization APIs.
+pub struct GdalXmlNode(NonNull<CPLXMLNode>);
+
+impl GdalXmlNode {
+    /// Create a Self from a raw pointer.
+    ///
+    /// # Safety
+    /// Caller is responsible for ensuring `ptr` is not null, and
+    /// ownership of `ptr` is properly transferred.
+    pub unsafe fn from_ptr(ptr: *mut CPLXMLNode) -> GdalXmlNode {
+        Self(NonNull::new_unchecked(ptr))
+    }
+
+    pub fn as_ptr(&self) -> *const CPLXMLNode {
+        self.0.as_ptr()
+    }
+
+    pub fn as_ptr_mut(&self) -> *mut CPLXMLNode {
+        self.0.as_ptr()
+    }
+}
+
+impl Clone for GdalXmlNode {
+    fn clone(&self) -> Self {
+        unsafe { GdalXmlNode::from_ptr(CPLCloneXMLTree(self.as_ptr())) }
+    }
+}
+
+impl Drop for GdalXmlNode {
+    fn drop(&mut self) {
+        unsafe { CPLDestroyXMLNode(self.0.as_mut()) };
+    }
+}
+
+impl FromStr for GdalXmlNode {
+    type Err = crate::errors::GdalError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let s2 = CString::new(s)?;
+        let c_xml = unsafe { CPLParseXMLString(s2.as_ptr()) };
+        if c_xml.is_null() {
+            Err(_last_cpl_err(CPLErr::CE_Failure))
+        } else {
+            Ok(unsafe { GdalXmlNode::from_ptr(c_xml) })
+        }
+    }
+}
+
+impl Display for GdalXmlNode {
```

Comment:
Should this also implement `ToString`?

Reply:
I've tended to conflate `Display` and `ToString` (since `Display` is `ToString`). Any advice on how to decide?

Reply:
Interesting... Docs say this:

<img width="443" alt="image" src="https://github.com/georust/gdal/assets/131013/9bd2157f-3ea8-41bf-a035-b0936658b3ac">